google/cloud/storage/client.py

```diff
@@ -122,6 +124,7 @@ def __init__(
             if client_options.api_endpoint:
                 api_endpoint = client_options.api_endpoint
                 kw_args["api_endpoint"] = api_endpoint
+                _override_private_url_variables(api_endpoint)
```

Comment:
Thanks for putting this together. I think the code needs to be refactored for maintainability but the this is a good start.

The URLs in URL Templates could be refactored to be pulled in from the client which is passed around the library.

For example:
https://github.com/googleapis/python-storage/blob/a50cdd145d39cb67e0d7021c376acbe376436178/google/cloud/storage/blob.py#L1020

Could be updated to:
```
_BASE_UPLOAD_TEMPLATE =  u"{hostname}/upload/storage/v1{bucket_path}/o?uploadType="
_MULTIPART_URL_TEMPLATE = _BASE_UPLOAD_TEMPLATE + u"multipart"
# truncated....
base_url = _MULTIPART_URL_TEMPLATE.format(hostname=storage_client._connection.API_BASE_URL,bucket_path=self.bucket.path)
```

WDYT?